[UnifiedPDF] [iOS] Support context menu when long pressing on links

whsieh opened the pull request:
#### 9cce2afd2b8d404682d457ca8b6eb5b519520101
<pre>
[UnifiedPDF] [iOS] Support context menu when long pressing on links
<a href="https://bugs.webkit.org/show_bug.cgi?id=284760">https://bugs.webkit.org/show_bug.cgi?id=284760</a>
<a href="https://rdar.apple.com/141555318">rdar://141555318</a>

Reviewed by Abrar Rahman Protyasha.

Add support for long-pressing on links in PDFs on iOS, when unified PDF is enabled. See below for
more details.

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:

Add an `isInPlugin` flag to position information, so that we can bail early if the user selects
&quot;Copy Link&quot; and handle the interaction without round-tripping back to the web process.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView actionSheetAssistant:performAction:]):
(-[WKContentView _tryToCopyLinkURLFromPlugin]):

Handle the Copy action immediately, in the case where the long-pressed &quot;element&quot; is a link
annotation inside a PDF plugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::linkURLAndBoundsAtPoint const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::linkURLAndBoundsAtPoint const):

Add a new method to retrieve the link URL and bounds at the given point in root view coordinates;
reimplement `highlightRectForTapAtPoint` on top of `linkURLAndBoundsAtPoint`. Note that I opted to
keep this separate method instead of removing it and just using `linkURLAndBoundsAtPoint` when
getting rects for the tap highlight, since we&apos;ll probably want to support other kinds of tappable
annotations (for which we want to show the tap highlight) in the future.

(WebKit::UnifiedPDFPlugin::highlightRectForTapAtPoint const):
(WebKit::UnifiedPDFPlugin::handleSyntheticClick):

Drive-by fix: allow shift-tap on links with an active selection to trigger navigation.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::linkURLAndBoundsAtPoint const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::positionInformation):

Populate `isLink`, `url`, `bounds`, and `isInPlugin` when long-pressing a link in a PDF.

Canonical link: <a href="https://commits.webkit.org/287912@main">https://commits.webkit.org/287912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcb0c90930ef6315f89faf710558690098b32336

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35247 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85833 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/32290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8645 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/32290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84373 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73995 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43765 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/472 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30748 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28738 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87268 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8534 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71781 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8715 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69825 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71014 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17677 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15069 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8496 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8332 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3894,22 +3894,39 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
-        return std::nullopt;
+        return { };
+
+    return { [annotation URL], pageToRootView([annotation bounds], [annotation page]) };
```
It's possible that the annotation is an in-document link, in which case `[annotation URL]` is `nil`.

aprotyas replied:
Hmm, it looks like `_tryToCopyLinkURLFromPlugin` does the right thing (bails when the URL is nil)

aprotyas replied:
But we probably want to change how we're calling this in `highlightRectForTapAtPoint()`, because we do a no-op for document destination links.

whsieh replied:
Ah, good point — we probably want to continue showing a tap highlight for in-document links then, but not a full context menu. I'll ensure that by changing `highlightRectForTapAtPoint` to:

```
std::optional<FloatRect> UnifiedPDFPlugin::highlightRectForTapAtPoint(FloatPoint pointInRootView) const
{
    // FIXME: We only support tapping on links at the moment. In the future, we might want to
    // support more types of annotations.
    auto [url, rect] = linkURLAndBoundsAtPoint(pointInRootView);
    if (rect.isEmpty())
        return std::nullopt;

    return rect;
}
```

whsieh committed:
[UnifiedPDF] [iOS] Support context menu when long pressing on links
https://bugs.webkit.org/show_bug.cgi?id=284760
rdar://141555318

Reviewed by Abrar Rahman Protyasha.

Add support for long-pressing on links in PDFs on iOS, when unified PDF is enabled. See below for
more details.

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:

Add an `isInPlugin` flag to position information, so that we can bail early if the user selects
"Copy Link" and handle the interaction without round-tripping back to the web process.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView actionSheetAssistant:performAction:]):
(-[WKContentView _tryToCopyLinkURLFromPlugin]):

Handle the Copy action immediately, in the case where the long-pressed "element" is a link
annotation inside a PDF plugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::linkURLAndBoundsAtPoint const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::linkURLAndBoundsAtPoint const):

Add a new method to retrieve the link URL and bounds at the given point in root view coordinates;
reimplement `highlightRectForTapAtPoint` on top of `linkURLAndBoundsAtPoint`. Note that I opted to
keep this separate method instead of removing it and just using `linkURLAndBoundsAtPoint` when
getting rects for the tap highlight, since we'll probably want to support other kinds of tappable
annotations (for which we want to show the tap highlight) in the future.

(WebKit::UnifiedPDFPlugin::highlightRectForTapAtPoint const):
(WebKit::UnifiedPDFPlugin::handleSyntheticClick):

Drive-by fix: allow shift-tap on links with an active selection to trigger navigation.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::linkURLAndBoundsAtPoint const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::positionInformation):

Populate `isLink`, `url`, `bounds`, and `isInPlugin` when long-pressing a link in a PDF.

Canonical link: https://commits.webkit.org/287912@main
